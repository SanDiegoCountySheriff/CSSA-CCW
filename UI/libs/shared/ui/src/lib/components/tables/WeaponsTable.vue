<!-- eslint-disable @intlify/vue-i18n/no-raw-text -->
<template>
  <v-container fluid>
    <v-data-table
      hide-default-footer
      :headers="headers"
      :items="weapons"
      mobile-breakpoint="800"
    >
      <template #top>
        <v-toolbar>
          <v-toolbar-title> Weapon Information </v-toolbar-title>
          <v-spacer></v-spacer>
          <v-btn
            @click="weaponDialog = true"
            color="primary"
            small
          >
            {{ $t('Add Weapon') }}
          </v-btn>
          <!-- <WeaponsDialog :item="{make: '', model: '', caliber: '', serialNumber: ''}" /> -->
        </v-toolbar>
      </template>
      <template #[`item.actions`]="{ item }">
        <v-tooltip
          top
          open-delay="500"
        >
          <template #activator="{ on, attrs }">
            <v-icon
              v-bind="attrs"
              @click="handleDelete(item)"
              v-on="on"
            >
              mdi-delete
            </v-icon>
          </template>
          <span>{{ $t('Delete item') }}</span>
        </v-tooltip>
        <v-icon
          @click="editWeapon(item)"
          color="primary"
          small
        >
          {{ $t('Edit Weapon') }}
        </v-icon>
        <!-- <WeaponsDialog
          :editing="true"
          :item="item"
        /> -->
      </template>
    </v-data-table>
<!-- research listeners for vue events, that will catch the event and re-emit it-->
    <WeaponsDialog
      v-model="weaponDialog"
      :item="editedWeapon"
      v-on="$listeners"
    />
  </v-container>
</template>

<script setup lang="ts">
import { WeaponInfoType } from '@shared-utils/types/defaultTypes'
import WeaponsDialog from '@shared-ui/components/dialogs/WeaponsDialog.vue'
import { ref } from 'vue'

interface IWeaponTableProps {
  weapons: Array<WeaponInfoType>
}

const emit = defineEmits(['delete', 'edit'])
const editedWeaponIndex = ref(-1)
const editedWeapon = ref({
  make: '',
  model: '',
  caliber: '',
  serialNumber: '',
})
const weaponDialog = ref(false)

const props = defineProps<IWeaponTableProps>()

const headers = [
  { text: 'Make', value: 'make' },
  { text: 'Model', value: 'model' },
  { text: 'Caliber', value: 'caliber' },
  { text: 'Serial Number', value: 'serialNumber' },
  { text: 'Actions', value: 'actions' },
]

function handleDelete(index) {
  emit('delete', index)
}

//how can i correctly implement handle edit here? what difference is this handleedit versus the one in weaponsdialog and workinfostep?
function handleEdit(index){
  editWeapon(index)
}

function editWeapon(item: WeaponInfoType) {
  // const item = props.weapons

  // like hair colors, get indexOf from the weapons array
  editedWeaponIndex.value = props.weapons.indexOf(item)
  editedWeapon.value.model = item.model
  editedWeapon.value.caliber = item.caliber
  editedWeapon.value.serialNumber = item.serialNumber
  weaponDialog.value = true
}
</script>

<style lang="scss" scoped>
.theme--dark.v-data-table {
  background: #303030;
}
</style>
