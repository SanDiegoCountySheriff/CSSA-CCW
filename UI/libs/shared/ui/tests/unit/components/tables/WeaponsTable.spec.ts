/* eslint-disable @typescript-eslint/ban-ts-comment */
import Vuetify from 'vuetify'
import WeaponsTable from '@shared-ui/components/tables/WeaponsTable.vue'
import { createLocalVue, mount } from '@vue/test-utils'

const localVue = createLocalVue()
const weapon = {
  make: 'walther',
  model: 'pps',
  caliber: '9mm',
  serialNumber: '12345',
}

describe('WeaponsTable', () => {
  let vuetify
  let wrapper
  const tMock = {
    $t: text => text,
  }

  beforeEach(() => {
    vuetify = new Vuetify()
    //@ts-ignore
    wrapper = mount(WeaponsTable, {
      localVue,
      vuetify,
      mocks: tMock,
      propsData: {
        weapons: [weapon],
      },
    })
  })

  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Should render the weapon correctly', () => {
    const tableData = wrapper.findAll('td')

    expect(tableData.at(0).element.textContent).toContain('walther')
    expect(tableData.at(1).element.textContent).toContain('pps')
    expect(tableData.at(2).element.textContent).toContain('9mm')
    expect(tableData.at(3).element.textContent).toContain('12345')
  })
})
