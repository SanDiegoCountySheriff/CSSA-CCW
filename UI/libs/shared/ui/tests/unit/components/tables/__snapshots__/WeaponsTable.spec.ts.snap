// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeaponsTable Should match the snapshot 1`] = `
<div class="container container--fluid">
  <div class="v-data-table text-left my-3 theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <thead>
          <tr>
            <th>
              Make
            </th>
            <th>
              Model
            </th>
            <th>
              Caliper
            </th>
            <th>
              Serial Number
            </th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              walther
            </td>
            <td>
              pps
            </td>
            <td>
              9mm
            </td>
            <td>
              12345
            </td>
            <td><button type="button" class="v-btn v-btn--text theme--light v-size--default error--text" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close-circle theme--light error--text"></i></span></button><span class="v-tooltip v-tooltip--bottom"><!----></span></td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
